feat: Adding fr compatibility to smt variables api

Sarkoxed opened the pull request:
This pr updates FFTerm and FFITerm classes. Also CMakeLists are changed a little. 

### FFTerm

- Added Fr compatibility. Now you can perform arithmetic operations with symbolic variables using `bb::fr` rather than strings or `FFTerm::Const`

### FFITerm

- Added `.mod()` method, that performs the operation of taking the remainder on terms. It leads to less constrained system of equations and helps the solver to work faster. However, now you have to do it manually.
- Added Fr compatiblitity(same here)
- Added xor, $\ge, \gt, \le, \lt$ operations
- Changed Constant variable initialization a little


Since you need these classes to behave similarly, they now both have identical methods. However, most of them are just placeholders in `FFTerm` class.


### Tests

Added simple unit tests for both ` FFTerm` and `FFIterm` to test that all the arithmetic operations are implemented properly.


## updt

### Terms
- changed division method to be more readable
- changed the hash value that is used as a dvision symbolic variable
  name to md5(Aztec)

- Removed the mod operations from inequalities methods in FFITerm

- Modified `==` and `!=` methods in FFITerm so now there won't be any
  redundant mod operations

Sarkoxed committed:
New more detailed CMakeLists

Sarkoxed committed:
Modified FFTerm and FFITerm classes

- Now they are compatible with Fr variables
- Changed FFIterm constant initialization

Sarkoxed committed:
Added tests for symbolic variables classes

Sarkoxed committed:
Removed xor test from ffiterm leaving only arithmetic + fr include

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/terms/ffiterm.cpp:
```diff
@@ -38,56 +38,52 @@ FFITerm::FFITerm(const std::string& t, Solver* slv, bool isconst, uint32_t base)
-            this->term = slv->s.mkInteger(tmp);
-        }
-        // this->term = slv->s.mkInteger(tmp); won't work for now since the assertion will definitely fail
+        std::string strvalue = slv->s.mkFiniteFieldElem(t, slv->fp, base).getFiniteFieldValue(); // TODO(alex): works for now
```
Could you add a more descriptive TODO? What do you wan to do here in the future?

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/terms/ffiterm.cpp:
```diff
@@ -102,13 +98,12 @@ void FFITerm::operator*=(const FFITerm& other)
-    cvc5::Term res = this->solver->s.mkConst(this->solver->s.getIntegerSort(),
-                                             "fe0f65a52067384116dc1137d798e0ca00a7ed46950e4eab7db51e08481535f2_div_" +
-                                                 std::string(*this) + "_" + std::string(other));
+    cvc5::Term res = Var("fe0f65a52067384116dc1137d798e0ca00a7ed46950e4eab7db51e08481535f2_div_" + std::string(*this) + "_" + std::string(other), this->solver).term;
```
What is this hash?

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/terms/ffiterm.cpp:
```diff
@@ -152,4 +146,64 @@ void FFITerm::operator!=(const FFITerm& other) const
+}
+
+void FFITerm::operator<(const bb::fr& other) const{
+    cvc5::Term lt = this->solver->s.mkTerm(cvc5::Kind::INTS_MODULUS, {this->term, this->modulus});
```
Why are you doing modular reduction in these?

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/terms/ffiterm.hpp:
```diff
@@ -11,39 +11,53 @@ using namespace smt_solver;
     static FFITerm Const(const std::string& val, Solver* slv, uint32_t base = 16);
 
+    explicit FFITerm(bb::fr value, Solver* s){
+        std::stringstream buf; // TODO(alex): looks bad. Would be great to create tostring() converter
```
Please add an issue

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/terms/ffiterm.test.cpp:
```diff
@@ -0,0 +1,101 @@
+
+using namespace smt_terms;
+
+TEST(integermod, addition)
```
could you change the name of the class of tests from integermod to FFITerm?

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/terms/ffterm.hpp:
```diff
@@ -17,27 +17,43 @@ class FFTerm {
     static FFTerm Const(const std::string& val, Solver* slv, uint32_t base = 16);
 
+    FFTerm(bb::fr value, Solver* s){
+        std::stringstream buf; // TODO(alex): looks bad. Would be great to create tostring() converter
```
Please create an issue

Rumata888 commented on barretenberg/cpp/src/barretenberg/smt_verification/terms/ffterm.test.cpp:
```diff
@@ -0,0 +1,95 @@
+
+using namespace smt_terms;
+
+TEST(finitefield, addition)
```
Could you rename from finitefield to FFTerm?

Sarkoxed committed:
Fixed FFITerm and FFITerm methods

- changed division method to be more readable
- changed the hash value that is used as a dvision symbolic variable
  name to md5(Aztec)

- Removed the mod operations from inequalities methods in FFITerm

- Modified `==` and `!=` methods in FFITerm so now there won't be any
  redundant mod operations

Sarkoxed committed:
Fix todos to be linked to issue

Sarkoxed committed:
Renamed the test classes

Rumata888 reviewed (approved):
LGTM

Sarkoxed committed:
Merge branch 'master' into as/smt-fr-compatiblity

Sarkoxed committed:
Merge branch 'master' into as/smt-fr-compatiblity

Sarkoxed committed:
Fixing a typo in readme.md

- FFTerm -> FFITerm

Sarkoxed committed:
Merge branch 'master' into as/smt-fr-compatiblity

Merging to merge

Sarkoxed committed:
Merge branch 'master' into as/smt-fr-compatiblity

Sarkoxed committed:
Merge branch 'master' into as/smt-fr-compatiblity

Rumata888 committed:
Merge branch 'master' into as/smt-fr-compatiblity

Rumata888 committed:
formatting fix
